pal/src/threadapi_pthreads_sl.c

```diff
@@ -0,0 +1,127 @@
+
+static void* ThreadWrapper(void* threadInstanceArg)
+{
+    THREAD_INSTANCE* threadInstance = (THREAD_INSTANCE*)threadInstanceArg;
+    int result = threadInstance->ThreadStartFunc(threadInstance->Arg);
+    return (void*)(intptr_t)result;
+}
+
+THREADAPI_RESULT ThreadAPI_Create(THREAD_HANDLE* threadHandle, THREAD_START_FUNC func, void* arg)
+{
+    THREADAPI_RESULT result;
+
+    if ((threadHandle == NULL) ||
+        (func == NULL))
+    {
+        result = THREADAPI_INVALID_ARG;
+        LogError("(result = %s)", ENUM_TO_STRING(THREADAPI_RESULT, result));
+    }
+    else
+    {
+        THREAD_INSTANCE* threadInstance = malloc(sizeof(THREAD_INSTANCE));
+        if (threadInstance == NULL)
+        {
+            result = THREADAPI_NO_MEMORY;
+            LogError("(result = %s)", ENUM_TO_STRING(THREADAPI_RESULT, result));
+        }
+        else
+        {
+            threadInstance->ThreadStartFunc = func;
+            threadInstance->Arg = arg;
+            int createResult = pthread_create(&threadInstance->Pthread_handle, NULL, ThreadWrapper, threadInstance);
+            switch (createResult)
+            {
+            default:
+                free(threadInstance);
+
+                result = THREADAPI_ERROR;
+                LogError("(result = %s)", ENUM_TO_STRING(THREADAPI_RESULT, result));
+                break;
+
+            case 0:
+                *threadHandle = threadInstance;
+                result = THREADAPI_OK;
+                break;
+
+            case EAGAIN:
+                free(threadInstance);
+
+                result = THREADAPI_NO_MEMORY;
+                LogError("(result = %s)", ENUM_TO_STRING(THREADAPI_RESULT, result));
+                break;
+            }
+        }
+    }
+
+    return result;
+}
+
+THREADAPI_RESULT ThreadAPI_Join(THREAD_HANDLE threadHandle, int* res)
+{
+    THREADAPI_RESULT result;
+
+    THREAD_INSTANCE* threadInstance = (THREAD_INSTANCE*)threadHandle;
+    if (threadInstance == NULL)
+    {
+        result = THREADAPI_INVALID_ARG;
+        LogError("(result = %s)", ENUM_TO_STRING(THREADAPI_RESULT, result));
+    }
+    else
+    {
+        void* threadResult;
+        if (pthread_join(threadInstance->Pthread_handle, &threadResult) != 0)
+        {
+            result = THREADAPI_ERROR;
+            LogError("(result = %s)", ENUM_TO_STRING(THREADAPI_RESULT, result));
+        }
+        else
+        {
+            if (res != NULL)
+            {
+                *res = (int)(intptr_t)threadResult;
+            }
+
+            result = THREADAPI_OK;
+        }
+
+        free(threadInstance);
+    }
+
+    return result;
+}
+
+void ThreadAPI_Exit(int res)
+{
+    pthread_exit((void*)(intptr_t)res);
+}
+
+void ThreadAPI_Sleep(unsigned int milliseconds)
+{
+    time_t seconds = milliseconds / 1000;
+    long nsRemainder = (milliseconds % 1000) * 1000000;
+    struct timespec timeToSleep = { seconds, nsRemainder };
+    clock_nanosleep(CLOCK_MONOTONIC, 0, &timeToSleep, NULL);
```

Comment:
>clock_nanosleep [](start = 4, length = 15)

`clock_nanosleep()` returns an error value, but this is not checked. We don't check it in our port in `threadapi_pthreads.c`, either; we cast the call to `(void)` to dismiss the error.

Reply:
I want to address the comment below but not sure how best to do it, so I will type here.

Originally we started by using threadapi_pthreads.c that was provided in the SDK. However, it defines _DEFAULT_SOURCE, which led to some compilation errors in GCC due to some pthread symbols being redefined as a result. In addition, we do not support nanosleep() in our posix layer, so we have to resort to overriding your implementation.